include/forcing/DataProvider.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef NGEN_DATAPROVIDER_HPP
+#define NGEN_DATAPROVIDER_HPP
+
+#include <string>
+
+
+namespace data_access
+{
+    /** An abstraction for classes that provider access to data.
+    *
+    * Data may be pre-provided from some external source, internally calculated by the implementing type, or some
+    * combination of both.
+    */
+
+    template <class data_type, class selection_type> class DataProvider
+    {
+        /** This class provides a generic interface to data services
+        *
+        */
+
+        public:
+
+        virtual ~DataProvider() = default;
+
+        /** Return the variables that are accessable by this data provider */
+
+        virtual const std::veector<std::string> get_avaliable_variable_names() = 0;
+
+        /** Return the first valid time for which data from the request variaable  can be requested */
+
+        virtual get_data_start_time(std::string var) = 0;
+
+        /** Return the last valid time for which data from the requested variablle can be requested */
+
+        virtual get_data_stop_time(std::string var) = 0;
+
+        /**
+         * Get the index of the data time step that contains the given point in time.
+         *
+         * An @ref std::out_of_range exception should be thrown if the time is not in any time step.
+         *
+         * @param epoch_time The point in time, as a seconds-based epoch time.
```

Comment:
What is the reference point for the epoch time?  Should we formalize that?

Reply:
In the case of NetCDF we can embed an epoch start time in the files. In other cases maybe set by ENV variables with a fallback to Unix standard epoch?